### ISSUE TITLE
IPv6: udhcpc/odhcp6c hard coded to request address, breaks DHCP-PD for some ISPs

### ISSUE THREAD

**Issue by xeth137** [RMerl/asuswrt-merlin#1517]
My ISP (Wave G) uses SLAAC for IPv6 address assignment and DHCP-PD for prefix assignment.  If a DHCPv6 solicit is sent upstream with both an address request (IA_NA) and a prefix request (IA_PD), the server responds with status "NOADDRS-AVAIL", presumably because IA_NA is not configured.  In other words, DHCP-PD does not work for my ISP if address requests are sent in same solicit.

Looking through the code in udhcpc.c I see that it launches odhcp6c with the hardcoded option "-N try" which means it always sends an address request.  In my use case, this needs to be changed to "-N none" to avoid the issue.

My current workaround for this is use the wan-start user script to kill the odhcp6c instance started by udhcpc and start my own.  It would be nice if I can override the argument with an NVRAM flag or custom conf instead.

**TL/DR**: Need a way to configure odhcp6c to NOT send an address request (i.e. "-N none" instead of "-N try"), because it breaks IPv6 on some ISPs.  Please consider implementing this in 382 if not 380.

Firmware version 380.68_4, Model RT-AC56U

**Comment by xeth137** [comment]
Here's a wire capture of the current (problematic) behavior:

[wireshark.txt](https://github.com/RMerl/asuswrt-merlin/files/1432040/wireshark.txt)

**Comment by RMerl** [comment]
Can you provide a longer trace that would include the full dhcpv6 exchange?  I'll have someone take a look at them (since I don't know much about IPv6 myself, and can't test it either).

It's possible that the ISP isn't properly following RFCs, since -N try should normally be able to fallback if the request failed.

**Comment by xeth137** [comment]
Thanks for the quick reply.  It looks like the fallback code only handles noaddrsavail in a REPLY, whereas in my case that flag is in an ADVERTISE (afaik it goes solicit -> advertise -> request -> reply).

I don't know much about IPv6 either so I'm not sure what the RFC behavior should be.  I'll try to post a full trace later today, but basically it's going to be the solicit/advertise pair I posted repeated over and over.

**Comment by xeth137** [comment]
Here's a full trace of udp port 546/547 during bootup:

[vlan2_dhcpv6_anon.zip](https://github.com/RMerl/asuswrt-merlin/files/1432683/vlan2_dhcpv6_anon.zip)

I also dug through the IPv6 RFCs and it appears the server does not follow RFC 7550 (which was an update written specifically to address this issue), however that RFC also says clients should handle this legacy behavior for backwards compatibility.

https://tools.ietf.org/html/rfc7550#section-4.1

**Comment by RMerl** [comment]
Can you test using this version?

http://www.mediafire.com/file/9cerclmabtpcfke/odhcp6c-test1.zip

Someone sent me a patch that might help make it more robust in such scenarios.  Test it by running it with "-N try".

My contact is currently busy with other things so he won't have the time to look at the trace for a few days.

**Comment by RMerl** [comment]
Ok, try this test2 build now:

http://www.mediafire.com/file/gqc49k8snik1f5q/odhcp6c-test2.zip

Test it by running it with "-N try" to see if it can properly fallback.

**Comment by xeth137** [comment]
running

```
odhcp6c -fve -R -s /tmp/dhcp6c -N try -c <SNIP> -PF 0:110f2 -r23 -r24 -r82 -r83 -k vlan2
```
results in syslog output of "invalid argument: 'F'"

...Taking away the F argument and everything works as expected.

```
odhcp6c -fve -R -s /tmp/dhcp6c -N try -c <SNIP> -PF 0:110f2 -r23 -r24 -r82 -r83 -k vlan2
```

It lo

But take away the F argument and everything works as expected.

```
odhcp6c -fve -R -s /tmp/dhcp6c -N try -c <SNIP> -PF 0:110f2 -r23 -r24 -r82 -r83 -k vlan2
```

Not using -F with the unpatched binary also fixes the issue.  It looks like force prefix flag prevents the -N try fallback code from working properly.

**Comment by themiron** [comment]
@xeth137 you're using wrong -F & -P options order, -FP is valid.

**Comment by xeth137** [comment]
My bad, using -FP with -N try does work with the patch.

**Comment by RMerl** [comment]
Thanks for testing.  I`ll keep the issue open for now, please confirm that it`s still working correctly after a day or two.

@themiron : I plan on merging the updated odhcp6c (not just this patch) with 380.69 and 382.2.

**Comment by themiron** [comment]
@RMerl great to know, thanks

**Comment by themiron** [comment]
@xeth137 yep, thanks for testing, please make sure it correctly renews the prefixes after a while (not sure what's your lease time)

**Comment by xeth137** [comment]
t1 is 900 and t2 is 1440, and it's been up for 6 hours now, looking good

**Comment by xeth137** [comment]
Uptime of 47 hours with no issues.  Thanks to everyone for the fix!
